stdlib/public/core/FloatingPoint.swift.gyb

```diff
@@ -2374,6 +2374,111 @@ extension BinaryFloatingPoint {
   */
 }
 
+% for Range in ['Range', 'ClosedRange']:
+%   exampleRange = '10.0..<20.0' if Range == 'Range' else '10.0...20.0'
+extension BinaryFloatingPoint
+where Self.RawSignificand : FixedWidthInteger,
+      Self.RawSignificand.Stride : SignedInteger & FixedWidthInteger,
+      Self.RawSignificand.Magnitude : UnsignedInteger {
+
+  /// Returns a random value within the specified range, using the given
+  /// generator as a source for randomness.
+  ///
+  /// Use this method to generate a floating-point value within a specific
+  /// range when you are using a custom random number generator. This example
+  /// creates three new values in the range `${exampleRange}`.
+  ///
+  ///     for _ in 1...3 {
+  ///         print(Double.random(in: ${exampleRange}, using: &myGenerator))
+  ///     }
+  ///     // Prints "18.1900709259179"
+  ///     // Prints "14.2286325689993"
+  ///     // Prints "13.1485686260762"
+  ///
+  /// The `random(in:using:)` static method chooses a random value from a
+  /// continuous uniform distribution in `range`, and then converts that value
+  /// to the nearest representable value in this type. Depending on the size and
+  /// span of `range`, some concrete values may be represented more frequently
+  /// than others.
+  ///
+  /// - Parameters:
+  ///   - range: The range in which to create a random value.
+%   if Range == 'Range':
+  ///     `range` must not be empty.
+%   end
+  ///   - generator: The random number generator to use when creating the
+  ///     new random value.
+  /// - Returns: A random value within the bounds of `range`.
+  @inlinable
+  public static func random<T: RandomNumberGenerator>(
+    in range: ${Range}<Self>,
+    using generator: inout T
+  ) -> Self {
```

Comment:
Here's another attempt at removing GYB code.

```swift
extension BinaryFloatingPoint where ... {

  public static func random<T: RandomNumberGenerator>(
    in range: ${Range}<Self>,
    using generator: inout T
  ) -> Self {
    _precondition(
      !range.isEmpty,
      "Can't get random value with an empty range"
    )
    let unitRandom = _unitRandom(using: &generator)
    let lowerBound = range.lowerBound
    let upperBound = range.contains(range.upperBound)
      ? range.upperBound
      : range.upperBound.nextDown
    return (upperBound - lowerBound) * unitRandom + lowerBound
  }
}
```

The internal API would be outside the `% for Range in ['Range', 'ClosedRange']:` block.

```swift
extension BinaryFloatingPoint where ... {

  // Returns a random value in the `0.0 ... 1.0` unit interval,
  // using raw values which can be represented exactly.
  internal static func _unitRandom<T: RandomNumberGenerator>(
    using generator: inout T
  ) -> Self {
    let max = (significandBitCount == RawSignificand.bitWidth - 1)
      ? RawSignificand.max
      : (1 as RawSignificand) << (significandBitCount + 1)
    let raw = RawSignificand.random(in: 0 ... max, using: &generator)
    return Self(exactly: raw)! / Self(exactly: max)!
  }
}
```

The [`significandBitCount`](https://developer.apple.com/documentation/swift/binaryfloatingpoint/1846714-significandbitcount) can be `Int.max` for extensible floating-point types. I'm not sure if we need to worry about that.

Reply:
I think `1.0 / ulpOfOne` gives the maximum precision needed.

```diff
-    let max = (significandBitCount == RawSignificand.bitWidth - 1)
-      ? RawSignificand.max
-      : (1 as RawSignificand) << (significandBitCount + 1)
+    let max = RawSignificand(1.0 / ulpOfOne)
```

Reply:
```swift
let upperBound = range.contains(range.upperBound)
      ? range.upperBound
      : range.upperBound.nextDown
    return (upperBound - lowerBound) * unitRandom + lowerBound
```

I'm not sure that logic works. Given `0.0 ..< 2.0`, `upperBound = 1` so `(1 - 0) * unit + 0` will only produce values within `0.0 ... 1.0` whereas we need values from `0.0 ..< 2.0`.

Reply:
@Azoy, `range.contains(range.upperBound)` is testing for a closed range; and `range.upperBound.nextDown` converts a half-open range into a closed range.

Given `0.0 ..< 2.0`, the local `upperBound` and the delta will be `1.9999999999999998`; I've tested this and it appears to work as intended.

Reply:
Oh my apologies, I assumed `nextDown` would return the next integer down.

Reply:
The non-branching version would be:

```swift
extension BinaryFloatingPoint where ... {

  public static func random<T: RandomNumberGenerator>(
    in range: ${Range}<Self>,
    using generator: inout T
  ) -> Self {
    _precondition(
      !range.isEmpty,
      "Can't get random value with an empty range"
    )
    let unitRandom = _unitRandom(using: &generator)
    let lowerBound = range.lowerBound
%   if 'Closed' in Range:
    let upperBound = range.upperBound
%   else:
    let upperBound = range.upperBound.nextDown
%   end
    return (upperBound - lowerBound) * unitRandom + lowerBound
  }
}
```